modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py

```diff
@@ -78,3 +78,139 @@ def _categories(self):  # noqa: GL08
+
+    Parameters
+    ----------
+    join_type : str
+        Join type.
+
+    Returns
+    -------
+    None
+    """
+    if join_type not in ("inner", "left"):
+        raise NotImplementedError(
+            f"{join_type} join is not supported by the HDK engine"
+        )
+
+
+def check_cols_to_join(what, df, col_names):
+    """
+    Check the frame columns.
+
+    Check if the frame (`df`) has the specified columns (`col_names`). The names referring to
+    the index columns are replaced with the actual index column names.
+
+    Parameters
+    ----------
+    what : str
+        Attribute name.
+    df : HdkOnNativeDataframe
+        The dataframe.
+    col_names : list of str
+        The column names to check.
+
+    Returns
+    -------
+    Tuple[HdkOnNativeDataframe, list]
+        The aligned data frame and column names.
+    """
+    cols = df.columns
+    new_col_names = col_names
+    for i, col in enumerate(col_names):
+        if col in cols:
+            continue
+        new_name = None
+        if df._index_cols is not None:
+            for c in df._index_cols:
+                if col == df._index_name(c):
+                    new_name = c
+                    break
+        elif df._index_cache is not None:
+            new_name = f"__index__{0}_{col}"
+            df = df._maybe_materialize_rowid()
+        if new_name is None:
+            raise ValueError(f"'{what}' references unknown column {col}")
+        if new_col_names is col_names:
+            # We are replacing the index names in the original list,
+            # but creating a copy.
+            new_col_names = col_names.copy()
+        new_col_names[i] = new_name
+    return df, new_col_names
+
+
+def join_by_index(
+    left,
+    right,
+    how,
+    left_on,
+    right_on,
+    sort,
+    suffixes,
+):  # noqa: GL08
+    def to_empty_pandas_df(df):
+        idx = df._index_cache
+        if idx is not None:
+            idx = idx[0:1]
+        elif df._index_cols is not None:
+            if len(df._index_cols) > 1:
+                arrays = [[i] for i in range(len(df._index_cols))]
+                names = [df._index_name(n) for n in df._index_cols]
+                idx = pandas.MultiIndex.from_arrays(arrays, names=names)
+            else:
+                idx = pandas.Index(name=df._index_name(df._index_cols[0]))
+        return pandas.DataFrame(columns=df.columns, index=idx)
+
+    new_dtypes = []
+    exprs = OrderedDict()
+    merged = to_empty_pandas_df(left).merge(
+        to_empty_pandas_df(right),
+        how=how,
+        left_on=left_on,
+        right_on=right_on,
+        sort=sort,
+        suffixes=suffixes,
+    )
+
+    if len(merged.index.names) == 1 and (merged.index.names[0] is None):
+        index_cols = None
+    else:
+        index_cols = left._mangle_index_names(merged.index.names)
+        for orig_name, mangled_name in zip(merged.index.names, index_cols):
+            df = left if mangled_name in left._dtypes else right
```

Comment:
could you please put a comment here with an explanation from [this comment](https://github.com/modin-project/modin/pull/5578#discussion_r1143912716)